pkg/controller/directvolumemigration/stunnel.go

```diff
@@ -0,0 +1,495 @@
+		var clientTpl bytes.Buffer
+		var destTpl bytes.Buffer
+		clientTemp, err := template.New("config").Parse(stunnelClientConfigTemplate)
+		if err != nil {
+			return err
+		}
+		destTemp, err := template.New("config").Parse(stunnelDestinationConfigTemplate)
+		if err != nil {
+			return err
+		}
+
+		// Execute templates
+		err = clientTemp.Execute(&clientTpl, stunnelConf)
+		if err != nil {
+			return err
+		}
+		err = destTemp.Execute(&destTpl, stunnelConf)
+		if err != nil {
+			return err
+		}
+
+		// Generate configmaps
+		clientConfigMap := corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: ns,
+				Name:      "directvolumemigration-stunnel-config",
+				Labels: map[string]string{
+					"app": "directvolumemigration-rsync-transfer",
+				},
+			},
+		}
+		err = yaml.Unmarshal(clientTpl.Bytes(), &clientConfigMap)
+		if err != nil {
+			return err
+		}
+
+		destConfigMap := corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: ns,
+				Name:      "directvolumemigration-stunnel-config",
+				Labels: map[string]string{
+					"app": "directvolumemigration-rsync-transfer",
+				},
+			},
+		}
+		err = yaml.Unmarshal(destTpl.Bytes(), &destConfigMap)
+		if err != nil {
+			return err
+		}
+
+		// Create configmaps on source + dest
+		err = srcClient.Create(context.TODO(), &clientConfigMap)
+		if k8serror.IsAlreadyExists(err) {
+			t.Log.Info("Configmap already exists on destination", "namespace", clientConfigMap.Namespace)
+		} else if err != nil {
+			return err
+		}
+
+		err = destClient.Create(context.TODO(), &destConfigMap)
+		if k8serror.IsAlreadyExists(err) {
+			t.Log.Info("Configmap already exists on destination", "namespace", destConfigMap.Namespace)
+		} else if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (t *Task) setupCerts() error {
+	// Get client for source
+	srcClient, err := t.getSourceClient()
+	if err != nil {
+		return err
+	}
+	// Get client for destination
+	destClient, err := t.getDestinationClient()
+	if err != nil {
+		return err
+	}
+
+	// steps
+	// 1. Generate CA cert
+	// 2. Loop through all namespace generating new certs for each namespace
+	// 3. Create secret in src+destination namespaces containing each cert
+	// 4. Rsync client+transfer pods mount certs from secret
+
+	// Skip CAbundle generation if configmap already exists
+	// TODO: Need to handle case where configmap gets deleted and 2 versions of
+	// CA bundle exist
+
+	caPrivKey, err := rsa.GenerateKey(rand.Reader, 4096)
+	if err != nil {
+		return err
+	}
+
+	subj := pkix.Name{
+		CommonName:         "openshift.io",
+		Country:            []string{"US"},
+		Province:           []string{"NC"},
+		Locality:           []string{"RDU"},
+		Organization:       []string{"Migration Engineering"},
+		OrganizationalUnit: []string{"Engineering"},
+	}
+
+	certTemp := x509.Certificate{
+		SerialNumber:          big.NewInt(2020),
+		Subject:               subj,
+		NotBefore:             time.Now(),
```

Comment:
hrm this needs to be reviewed carefully. The source, destination, and controller clusters can have different times and can trip us in ways that would cause a lot of pain

Reply:
Hmm... hadn't considered that. Let's chat about this.